### PR TITLE
adds generic neural net class

### PR DESCRIPTION
**Pull request by joeyklee**
This is a "best of both worlds" between @shiffman's example and @lydiajessup 's example. I think it is working!

### CODE REVIEW

File: NeuralNet.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+            const inputTensor = tf.tensor(inputs); 
+            const labelTensor = tf.oneHot(tf.tensor1d(labels, 'int32'), this.config.NO_VAL );
+
+            //Step 3. Normalize the data to the range 0 - 1 using min-max scaling
+            const inputMax = inputTensor.max();
+            const inputMin = inputTensor.min();  
+            const labelMax = labelTensor.max();
+            const labelMin = labelTensor.min();
+
+            const normalizedInputs = inputTensor.sub(inputMin).div(inputMax.sub(inputMin)).flatten().reshape([ data.length, this.config.inputs]);
```
Comment (joeyklee) [inline NeuralNet.js:116]: This is kind of a funky maneuver, but it is a way to correctly normalize the values and reshape the data to fit the model input shape.

Comment (joeyklee) [reply NeuralNet.js:116]: We should note that this only works for numeric data and not strings.

File: NeuralNet.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
+class NeuralNet {
+    constructor(options) {
```
Comment (shiffman) [inline NeuralNet.js:2]: (This is really a comment re: https://github.com/ml5js/ml5-library/pull/485/ but putting here in terms of thinking this all through.) For simple cases, I'm thinking about supporting passing in just number of inputs, number of outputs:

```javascript
let neuralnet = new NeuralNet(3, 2);
```

File: NeuralNet.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+        const hidden = tf.layers.dense({
+            units: 16,
+            inputShape: [this.config.inputs],
+            activation: 'relu'
+        });
+        
+        const output = tf.layers.dense({
+            units: this.config.outputs,
+            activation: 'softmax'
```
Comment (shiffman) [inline NeuralNet.js:30]: `softmax` wouldn't be the default, but only if specified as a "classification" task.

File: NeuralNet.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+
+        if (this.config.debug) {
+            tfvis.show.modelSummary({
+                name: 'Model Summary'
+            }, this.model);
+        }
+    }
+
+    async loadData(DATA_URL, dataType) {
+        const outputLabel = this.config.outputLabels[0];
```
Comment (shiffman) [inline NeuralNet.js:52]: How we handle the user specifying what is a "label" in the data for both inputs and outputs is tricky.

File: NeuralNet.js (hunk 1 above)
Comment (shiffman) [reply NeuralNet.js:116]: I wonder if we should expose a function like `normalize()` and have the user explicitly call it. I'm still confused about what makes the most sense in terms of the `NeuralNet` class having data functionality built in or that happens elsewhere and the user just passes in `xs` and `ys` to `train()`.

File: sketch.js [hunk 5]
```diff
@@ -1,158 +1,91 @@
...
+
+let nn;
+
+let data;
+const neuralNetOptions = {
+  inputs: 3,
+  outputs: 2,
+  debug: true,
+  outputLabels: ['survived'],
+  inputLabels: ['age', 'fare', 'is_female'],
```
Comment (shiffman) [inline sketch.js:23]: this is really great having the labels specified this way!

**Comment by shiffman** [comment]
@joeyklee this is just some stream of consciousness thoughts as I look through no need to respond or make changes based on them as we can move to https://github.com/ml5js/ml5-library/pull/485/ and work on the example for [Intro to ML](https://github.com/ml5js/Intro-ML-Arts-IMA) there!

**Review by shiffman — COMMENTED** [review]
@joeyklee So wonderful thank you for synthesizing so many disparate ideas! This review has some stream of consciousness thoughts as I looked through this code! No need to respond or make changes based on them as we can move to https://github.com/ml5js/ml5-library/pull/485/ and work on the example for [Intro to ML](https://github.com/ml5js/Intro-ML-Arts-IMA) there!

File: NeuralNet.js (hunk 2 above)
Comment (joeyklee) [reply NeuralNet.js:2]: I like making that explicit as well. We can do something similar to how we handle the other examples: e.g. `inputsOrOptions, outputsOrOptions`

File: NeuralNet.js (hunk 3 above)
Comment (joeyklee) [reply NeuralNet.js:30]: Oh! yes, maybe something then like:

```
new ml5.neuralNetwork(3, 2, 'regression') 
new ml5.neuralNetwork(3, 2, 'classification') 
```

then our activation function will change accordingly

File: sketch.js (hunk 5 above)
Comment (shiffman) [reply sketch.js:23]: Oh, actually I'm confused. We probably shouldn't use the term "labels"  as we're using this term for "class" elsewhere in the library. This is really the "field name" or "data heading", hmmmmm. And these are really part of the data and not the neural network itself. . .leading me again towards something like `ml5.data()`, hmmmm. For now maybe just:

```javascript
  outputs: ['survived'],
  inputs: ['age', 'fare', 'is_female']
```

File: NeuralNet.js (hunk 4 above)
Comment (joeyklee) [reply NeuralNet.js:52]: Totally. We'll have to put some more thoughts on this. Maybe we create a hashmap that stores the key:values ?

File: sketch.js (hunk 5 above)
Comment (joeyklee) [reply sketch.js:23]: This also allows users to load up data w/ extraneous fields.

File: NeuralNet.js (hunk 3 above)
Comment (shiffman) [reply NeuralNet.js:30]: Yes! And oddly enough I think I would default to regression, having the user specify only classification. Like a vanilla neural network natively is just going to give numbers, not output labels. (Most of these comments are about me thinking ahead to my interactive training a la wekinator examples).

Comment (joeyklee) [reply NeuralNet.js:30]: Yes, I also think this makes sense. I also think about the Nature of Code behaviors which also operate this way.

Comment (shiffman) [reply NeuralNet.js:30]: Indeed! I am going to work on revising Chapter 10 this week and have it as reading for the course next week!

Comment (joeyklee) [reply NeuralNet.js:30]: Awesome! Ok, I will ping you as I make updates on the neuralnet class!
